### ISSUE TITLE
Edit Correspondence patterns not functional

### ISSUE THREAD

**Issue by FredericBlum** [digling/edictor#223]
@LinguList I fail to access the correspondence patterns in the current edictor setup. If I select 'Edit Correspondence Patterns', I get a non-reactive window. When I click 'submit', nothing appears in my comand line running the server. When I open it, the following line appears:

```127.0.0.1 - - [02/Oct/2024 14:53:52] "GET /panels/patterns.html HTTP/1.1" 200 -```

![Screenshot 2024-10-02 at 14 54 58](https://github.com/user-attachments/assets/490492e0-bd96-44b1-a193-855d8ac45a64)

If I select 'Compute > Correspondences', suddenly the window with the correspondence patterns open! However, I am then not able to interact with the patterns (e.g. select different ones) because the computation is stuck loading, indicated by the spinning whirl. 

![Screenshot 2024-10-02 at 14 56 03](https://github.com/user-attachments/assets/b2dea5c5-4f5a-4ebe-a141-5c34db3b6d21)

In short, what am i doing wrong to access the corresponce patterns in my data?> Again, tested with migliazzayanomamic.

**Comment by LinguList** [comment]
Well, there are a few things to be kept in mind here: 

1. manual editing of correspondence patterns requires a column PATTERNS is that in your data?
2. be careful in mixing automatic (overwrites) and manual (does only modify parts) pattern editing

**Comment by LinguList** [comment]
So you miss the column "PATTERNS".

**Comment by LinguList** [comment]
Add them by modivying this line int he config:

```
          "columns": "DOCULECT|CONCEPT|VALUE|FORM|TOKENS|COGID|COGIDS|ROOTIDS|ALIGNMENT|MORPHEMES|PATTERNS|NOTE"
```

**Comment by LinguList** [comment]
Then, the important question: do you want to compute patterns on partial or full cognates?

**Comment by LinguList** [comment]
Computing patterns on the remote server is a bit of a thing, because you have many data points that will be modified, it may take some time, I rather do not do so now, I just use your modified config, and download the data, and open from there now.

**Comment by LinguList** [comment]
Luckily, we can save files now locally (it saves in the folder where you opened edictor). You restart the server, you also see the file under FILES (good for testing!)

**Comment by LinguList** [comment]
I open the file now, then I go to CONFIGURATION -> SETTINGS and then I set the colexification mode to FULL COGNATES (it is already there).

**Comment by LinguList** [comment]
Now all patterns HAVE been computed, as you can see in the column PATTERNS.

**Comment by LinguList** [comment]
This works all flawless, I can edit the patterns now.

**Comment by LinguList** [comment]
I just pushed the example file to the edictor-remote folder in the main branch of migliazzayanomamic.

**Comment by LinguList** [comment]
edictor.org has some help on this, but it is not extensive. Our tests, however, which run these cases, have the workflow covered. You must first add PATTERNS by going to add column, inserting PATTERNS, then edit the SETTINGS, and then compute the first time the patterns, since if you don't compute, you get no patterns to be manipulated.

**Comment by FredericBlum** [comment]
Thanks for the walkthrough. One issue remains, however. I can now access the column 'PATTERNS', but when I reload the dataset and select 'Edit Correspondence Patterns', the window remains empty and unresponsive. I have to go throught he computation again to access the pattern themselves.

![Uploading Screenshot from 2024-10-04 12-21-22.png…]()

**Comment by LinguList** [comment]
Did you also test with the file itself?

**Comment by LinguList** [comment]
![DeepinBildschirmfoto_Bereich auswählen_20241004223814](https://github.com/user-attachments/assets/8a63763d-cd91-4c19-9317-e7861e03b97c)

**Comment by LinguList** [comment]
Do you see those patterns in the data, @FredericBlum ? The are wrong. You must have at least a 0 for each sound in the alignment. Dunno, how they were introduced, you have empty patterns as well.

**Comment by LinguList** [comment]
The problem is, however deeper.

**Comment by LinguList** [comment]
But now check the sqlite version. (filter by tokens = +)

![DeepinBildschirmfoto_Bereich auswählen_20241004225124](https://github.com/user-attachments/assets/c738ce6c-0306-4d42-a3f5-8c06e57fc0e6)

against the text file

![DeepinBildschirmfoto_Bereich auswählen_20241004225136](https://github.com/user-attachments/assets/33e4cae5-6366-4da1-b812-71ee65ef8aa6)

**Comment by LinguList** [comment]
You have many + chars for the cognate sets that should not be there. I think it is that the server was not properly uploaded, or similar Since you edit a bunch of data at once, and this may not make it all to the server (my wild guess now).

**Comment by LinguList** [comment]
Note that the patterns show the index of the ALIGNMENT but the ALIGNMENT in its TRIMMED form. So if you re-trim the data, the Patterns must be checked.

**Comment by LinguList** [comment]
This can be done manually, but it is tedious.

**Comment by LinguList** [comment]
I just used the debugger to check if I can repair the data. But it is not possible. The ways to proceed:

1. make a fix in edictor that checks and ignores all sequences that have the problem (if a cognate set has conflicting pattern data), (partly handled, but obviously not good enough)
2. for the case at hand, correct the data based on the derived file, and use that instead of the online version (you could recreate the sqlite with lexibase)

**Comment by FredericBlum** [comment]
Ok, but the pattern data was never manipulated manually. Doesn't this mean that there is some problem with the online-computation of patterns?

**Comment by FredericBlum** [comment]
There seem to be a number of problems distinct, but related problems:

- Manually entered datapoints do not receive pattern ID's after alignment, needs manual assignment and change
- Compute Correspondence pattern does not seem to be able to work with "+' in Full Colexification mode, hence the problems in your screenshots
- Editing the alignments manually (e.g. removing '+') bricks the server when a column is selected for Correspondence Patterns - if empty, modifying the alignments works fine. But the patterns never get updated.

So right now, it seems like the only solution is to re-do the correspondence patterns locally, and reupload, as you suggested. And then keep doing manual modifications of the Pattern ID column.

**Comment by FredericBlum** [comment]
It also seems like deleting data does not really work. I have removed two empty rows again and again, but they keep reappearing.

**Comment by FredericBlum** [comment]
One example of the problem with the comptuation online:

'p i h i + p i h k ɨ' - as alignment, nothing trimmed - becomes '608 155 145 153 + 0 0 0 + 0 0' - everything after '+' just gets turned to 0.

**Comment by FredericBlum** [comment]
In the online version, I have removed all '+' from the forms and alignments, except for the singletons. The patterns do not have anything in the form of '+ 0 0 0 0' anymore. But the Edit Correspondence Patterns button is still unresponsive when submitting if I do not re-compute the patterns before. Maybe we could also hgave a look at this in the Oberseminar? I feel like I still didn't understand some parts that I should have.

**Comment by LinguList** [comment]
EDICTOR as a default assigns each segment the pattern 0, unless it is a plus. We assume that plus is only used when working with COGIDS, but that's not the case for your dataset.

**Comment by LinguList** [comment]
The way to go here seems to be to get back to a version on a text file and to experiment, instead of the server version. This is easy to handle, and we can verify that deleting all plusses will get rid of these problems. But the general pattern problem persists, and we must find ways to handle it in a workflow. 

- we want to edit patterns and modify alignments at the same time, but that will shift all patterns, it is very difficult to program code that would track the original pattern assignment when you modify an alignment
- so we may want to mix automated and non-automated assignment, but this is also difficult to do, of course
- and we have the server problem: if you compute lime 2000 items, they must be SEND to the SQLITE on the server, but this may cause problems with the way that JavaScript makes these updates in an asynchronous fashion

**Comment by LinguList** [comment]
Ah, I just saw, sorry, I misread what you wrote. So we can verify that when working with a file and not with the server, the editing WORKS. This seems important to me.

**Comment by LinguList** [comment]
If that's the case, it is most likely that such long updates are not possible (or I may have even MISSED code updates on the server, so they never GET there).

**Comment by LinguList** [comment]
Assuming you work alone with the data now and want to explore just the correspondences, my recommendation is to switch to file edit-mode. You put the file next to the config, open edictor, and you find it in the tab FILES. Clicking on the right-most save button will store the file and leave a back-up of the previous version. So you have full account of what you edited.

**Comment by LinguList** [comment]
When I find time, we then resume the SQLITE issue and we discuss also these details in the OS!

**Comment by FredericBlum** [comment]
I will get to this next week, I hope. I am surprised that this doesn't work, because it was a standard part of the workflow in the old EDICTOR. E.g. in girardpanotakanan and valzarpanotakana, we have full cognacy and morpheme segmentation, which did not disturb the method. But it's good to know that we should avoid this. Of course, the cleaner implementation would be partial cognacy anyway, but it's unclear if we can invest the time necessary for this.

**Comment by LinguList** [comment]
It was always standard to COMPUTE, never to ANNOTATE. And now, you compute and send to the server, this was also never done.

**Comment by LinguList** [comment]
And it works -- as far as I have been confirming with your data -- if you use the file, not the server.

**Comment by LinguList** [comment]
Is this AFTER my update? I only saw your post now, but I think I addressed the pattern question in #225, does the patch there help here?

**Comment by FredericBlum** [comment]
This was before, but the problem persists. It seems like there is a CoPaR error involved:

```
16 C V C V | m a !S/s i ( - - ) | m a !S/s i
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 16
17 C V C V C V | m a s i ( t a ) | m a s i t a
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 17
18 C V | p o ( - - ) | p o
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 18
19 C V C V | p o ( k o ) | p o k o
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 19
20 C V C V | ʃ i r o ( - - ) | ʃ i r o
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 20
21 V C V | - i ɾ o ( - - ) | i ɾ o
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 21
22 C V C V C V | s i r o ( m a ) | s i r o m a
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 22
50 C V C V | t a w a ( - - ) | t a w a
2024-11-01 10:24:24,019 [WARNING] alignment and structure do not match in 50
51 C V C V C V | ʃ a ɾ a ( k a ) | ʃ a ɾ a k a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 51
52 C V C V |  | p o k o
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 52
53 C V C V |  | p u s i
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 53
59 V C V C |  | o ɾ ɨ k
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 59
60 C V C V | ʔ a ʃ a ( - - ) | ʔ a ʃ a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 60
61 C V | j a - - ( - - ) | j a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 61
62 C V C V C V | j a ɾ a ( k a ) | j a ɾ a k a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 62
75 C V C V | k u k u ( - - ) | k u k u
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 75
76 V C V C V | - u k u ( k a ) | u k u k a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 76
86 C V | t u ( - - - - ) | t u
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 86
87 C V C V C V | t u ( l ɨ t ɨ ) | t u l ɨ t ɨ
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 87
88 C V C V | ( - - ) b i !N/n a | b i !N/n a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 88
89 C V C V C V | ( k o ) p e n a | k o p e n a
2024-11-01 10:24:24,020 [WARNING] alignment and structure do not match in 89
```

These are of two types:

- Isolates get initiated without any alignment. If I add the 0's manually, they are removed from the errors. But for larger datasets, this would not be feasible
- Ignored elements also throw errors. If I do not trim any part of the alignment, the entry is also not listed on the errors. But it seems like the current implementation has problems correctly identifying the brackets.

**Comment by LinguList** [comment]
Isolates are singletons, right? This is of course expected. The problem you see there in COPAR is due to the clash of trimming and a structure column that is not trimmed. But this is something that COPAR usually does automatically. It would be good to check against the method that is NOT using lingpy/lingrex, to make sure this is an error in lingrex or in edictor.

**Comment by FredericBlum** [comment]
I think we had similar problems previously when working on the lingreg experiments. See discussion here: https://github.com/lingpy/lingrex/issues/41

This led us to write some code for reducing the alignments: https://github.com/pano-takanan-history/lingreg/blob/shuffle/example/cogids_reduce.py

What I was surprised by is that those problems did not occur with the old Edictor. Singletons where no problem for the computation of alignments, neither was trimming. Or if it was, it was not made explicit/transparent.

Perhaps we should build a finer version of the cogids_reduce package, and build the correspondence patterns upon the reduced version of the alignments? And remove the singletons explicitly from the computation, but without throwing an error. The utility for this is already implemented in lingrex/util.py::(prep_wordlist(min_refs=2)), which filters wordlist for cognate set siwth minimum two members. (link here: https://github.com/lingpy/lingrex/blob/master/src/lingrex/util.py#L194)

When I computed correspondence patterns with trimmed data, I always had to reduce the alignments like this, taking out the trimmed part.

**Comment by LinguList** [comment]
> What I was surprised by is that those problems did not occur with the old Edictor. Singletons where no problem for the computation of alignments, neither was trimming. Or if it was, it was not made explicit/transparent.

The old edictor did not have global computation of alignments, though. I think we should start from a *fresh* dataset that we use to test through. We start from ONLY using a dataset in a file. If we confirm this works, we go to sqlite. Now, we mix up several different parts and it is hard for me to get at it, as we always have varying datasets I must install etc.

**Comment by LinguList** [comment]
We can quickly discuss this tomorrow, @FredericBlum

**Comment by FredericBlum** [comment]
Found two ways to break things:

- using COGIDS instead of COGID I receive an error message for two cases that involve '+'. This is probably not faithfully removed when computing the structure? Since the alignment is correct.

```
127.0.0.1 - - [21/Nov/2024 14:45:02] "GET /panels/ipatterns.html HTTP/1.1" 200 -
Loaded the CoPaR object.
70 C V + C V | - - k e | k e + m a
2024-11-21 14:45:02,896 [WARNING] alignment and structure do not match in 70
92 C V + C V | b u - - | b u + t ɨ
2024-11-21 14:45:02,896 [WARNING] alignment and structure do not match in 92
----------------------------------------
```

- trimming two sites in a cognate set, I receive two more error messages. This happens both when using COGID as well as using COGIDS:

```
127.0.0.1 - - [21/Nov/2024 14:45:02] "GET /panels/patterns.html HTTP/1.1" 200 -
127.0.0.1 - - [21/Nov/2024 14:46:06] "GET /css/images/ui-icons_222222_256x240.png HTTP/1.1" 200 -
Loaded the CoPaR object.
50 C V C V | t a w a ( - - ) | t a w a
2024-11-21 14:47:29,118 [WARNING] alignment and structure do not match in 50
51 C V C V C V | ʃ a ɾ a ( k a ) | ʃ a ɾ a k a
2024-11-21 14:47:29,118 [WARNING] alignment and structure do not match in 51
```

This is the above described case that the computation does not work for trimmed elements.

**Comment by FredericBlum** [comment]
I am not sure what changed, since I cannot find any changes in the commit history - but the problem does not seem to persist. Therefore, I will close this issue for now.
